File: 2021/day02/solution.py
from typing import Iterable, Tuple

def getLines(filename="input.txt") -> Iterable[str]:
    with open(filename) as f:
        line = f.readline()
        while line:
            yield line[:-1]
            line = f.readline()

def parseLine(line: str) -> Tuple[str, int]:
    parts = line.split(" ")
    return (parts[0], int(parts[1]))

def main():
    print(f"part1: {part1()}")
    print(f"part2: {part2()}")

def part1():
    lines = getLines()
    parsed = map(parseLine, lines)
    horizontal = 0
    vertical = 0
    for direction, amount in parsed:
        if direction == "forward":
            horizontal += amount
        elif direction == "down":
            vertical += amount
        else:
            vertical -= amount
    return f"horizontal={horizontal}, vertical={vertical}, product={horizontal * vertical}"

def part2():
    lines = getLines()
    parsed = map(parseLine, lines)
    horizontal = 0
    vertical = 0
    aim = 0
    for direction, amount in parsed:
        if direction == "forward":
            horizontal += amount
            vertical += aim * amount
        elif direction == "down":
            aim += amount
        else:
            aim -= amount
    return f"horizontal={horizontal}, vertical={vertical}, product={horizontal * vertical}"

if __name__ == "__main__":
    main()

File: 2021/day02/golang/solution.go
package main

import (
    "bufio"
    "log"
    "os"
    "strconv"
    "strings"
)


func check(err error) {
    if err != nil {
        log.Fatal(err)
    }
}

func getLines(path string) ([]string, error) {
    file, err := os.Open(path)
    defer file.Close()

    if err != nil {
        return nil, err
    }

    scanner := bufio.NewScanner(file)
    lines := make([]string, 0, 3000)
    for scanner.Scan() {
        line := scanner.Text()
        lines = append(lines, line)
    }

    return lines, nil
}

func main() {
    log.SetFlags(log.Lmicroseconds)
    part1("input.txt")
    part2("input.txt")
}

func part1(path string) {
    log.SetPrefix("part1: ")
    lines, err := getLines(path)
    check(err)
    horizontal := 0
    vertical := 0
    for _, line := range lines {
        parts := strings.Fields(line)
        direction := parts[0]
        amount, err := strconv.Atoi(parts[1])
        check(err)
        //log.Println(direction, amount)
        switch direction {
            case "forward": horizontal += amount
            case "down": vertical += amount
            case "up": vertical -= amount
        }
        //log.Printf("horizontal: %d, vertical: %d", horizontal, vertical)
    }
    log.Println("product:", horizontal * vertical)
}

func part2(path string) {
    log.SetPrefix("part2: ")
    lines, err := getLines(path)
    check(err)
    horizontal := 0
    vertical := 0
    aim := 0
    for _, line := range lines {
        parts := strings.Fields(line)
        direction := parts[0]
        amount, err := strconv.Atoi(parts[1])
        check(err)
        //log.Println(direction, amount)
        switch direction {
            case "forward": {
                horizontal += amount
                vertical += (aim * amount)
            }
            case "down": aim += amount
            case "up": aim -= amount
        }
        //log.Printf("horizontal: %d, vertical: %d", horizontal, vertical)
    }
    log.Println("product:", horizontal * vertical)
}
